### ISSUE TITLE
Question on SingleR method correlation by needing reference dataset stratification

### ISSUE THREAD

**Issue by mauritsunkel** [LTLA/SingleR#231]
Hi again, 

I was worried given an example case of using a reference dataset that contains e.g. 75% cell type A and 25% cell type B, any given cell during training would be more biased in classification towards cell type A, if not very clearly correlating with cell type B features. 

The base of my question is answered in the SingleR book chapter 1.2 Method description, which states:
"We define the per-label score as a fixed quantile (by default, 0.8) of the correlations across all samples with that label. This accounts for differences in the number of reference samples for each label, which interferes with simpler flavors of nearest neighbor classification; it also avoids penalizing classifications to heterogeneous labels by only requiring a good match to a minority of samples."

Now my question has become, how does "defining the per-label score as a fixed quantile (by default, 0.8) of the correlations across all samples with that label account for differences in the number of reference samples for each label"? Could you explain it to me by how it "differs from the simpler flavors of NN classification" or link me to where in the code/math this is taken into account? (I guess by getting this answered fundamentally I will also understand how it "avoids penalizing classifications to heterogeneous labels by only requiring a good match to a minority of samples.")

And thanks again, 
Maurits

**Comment by LTLA** [comment]
> Now my question has become, how does "defining the per-label score as a fixed quantile (by default, 0.8) of the correlations across all samples with that label account for differences in the number of reference samples for each label"?

Because the expectation of the score doesn't change with the number of reference samples.

Contrast this with a simpler nearest neighbor search, where you just define the label for each test cell as the most frequent label among its nearest neighboring reference cells. All things being equal, a more abundant label in the reference dataset will have a greater chance of obtaining a majority in the neighboring set, thus biasing classification towards it.

To demonstrate:

```r
set.seed(10)

# Make a reference with two labels with the same spread but different frequency.
label1 <- matrix(rnorm(50, sd = 0.1), ncol=2)
label1[,1] <- label1[,1] - 0.5
label2 <- matrix(rnorm(5000, sd =0.1), ncol=2)
label2[,1] <- label2[,1] + 0.5
ref <- rbind(label1, label2)
label <- rep(1:2, c(nrow(label1), nrow(label2)))

# Make a test dataset.
test <- matrix(runif(1000), ncol=2)
test[,1] <- test[,1] - 0.5
test[,2] <- test[,2] - 0.5

# Assign labels based on the majority of the nearest neighbors.
nn <- BiocNeighbors::queryKNN(X=ref, query=test, k=15)
assigned <- (rowMeans(nn$index <= nrow(label1)) < 0.5) + 1L

plot(ref[,1], ref[,2], col=label)
points(test[,1], test[,2], pch=16, col=assigned)
```

![Screen Shot 2022-09-09 at 8 33 51 AM](https://user-images.githubusercontent.com/8166669/189387658-305dc8d4-f6a5-4c07-97d9-d7ce1c15b910.png)

You can see that the boundary is not actually at the midpoint (0) of the two labels. Classification is biased towards the more abundant label in the reference, because the 15th neighbor is closer when you have more cells.

The quantile approach defends against this because it is based on the nearest proportion of neighbors, rather than the nearest number. In this manner, we adjust for the relative differences in abundance between labels.

Of course, a simpler solution would be to just classify based on the centroid of each cluster, but...

> I guess by getting this answered fundamentally I will also understand how it "avoids penalizing classifications to heterogeneous labels by only requiring a good match to a minority of samples."

What happens if the labels are highly heterogeneous? For example, anything labelled "neuron" is usually a big clump of distinct cell types. You don't want to classify against the centroid in those cases, because that would be some biologically irrelevant average mishmash of different types. Instead, we use the quantile because this allows us to focus on the nearest (proportion of) neighbors to the test cell. In this manner, we can still classify cells as neurons based on their closest neuronal subtype, without worrying about the relevance of the centroid.

Again:

```r
set.seed(20)

# Make a reference with two labels with different spread but the same frequency.
label1 <- matrix(rnorm(1000, sd = 0.01), ncol=2)
label1[,1] <- label1[,1] - 0.5
label2 <- matrix(rnorm(1000, sd =0.5), ncol=2)
label2[,1] <- label2[,1] + 0.5
ref <- rbind(label1, label2)
label <- rep(1:2, c(nrow(label1), nrow(label2)))

# Make a test dataset.
test <- matrix(runif(1000), ncol=2)
test[,1] <- test[,1] - 0.5
test[,2] <- test[,2] - 0.5

# Assign labels based on the closest mean.
means <- rbind(colMeans(label1), colMeans(label2))
nn <- BiocNeighbors::queryKNN(X=means, query=test, k=1)
assigned <- nn$index

plot(ref[,1], ref[,2], col=label, xlim=c(-1, 1), ylim=c(-0.5, 0.5))
points(test[,1], test[,2], pch=16, col=assigned)
```

![Screen Shot 2022-09-09 at 8 46 01 AM](https://user-images.githubusercontent.com/8166669/189389911-49e2e7d3-0984-42f1-857a-31414376ade5.png)


Arguably many of the points assigned to the black cluster would be more appropriately assigned to the red cluster.

**Comment by mauritsunkel** [comment]
Cheers LTLA, this explanation was more than clear and indeed the only discussion leftover is the accuracy/precision of the assigned points in the end, however, I now understand and agree with the underlying logic and theory of this choice, which is more important in my case, thanks again!
